### PR TITLE
nixos/postgresql: replace extraConfig option with settings option

### PR DESCRIPTION
**Pull request by aanderse**
###### Motivation for this change
https://github.com/NixOS/nixpkgs/pull/94231

ping @jappeace 

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/modules/services/databases/postgresql.nix [hunk 1]
```diff
@@ -212,10 +211,20 @@ in
...
+        default = {};
+        description = ''
+        '';
+        example = literalExample ''
+          {
+            log_connections = true;
+            log_statement = "'all'";
+            logging_collector = true
+            log_disconnections = true
+            log_destination = lib.mkForce "'syslog'";
```
Comment (jappeace) [inline nixos/modules/services/databases/postgresql.nix:225]: `${lib.mkForce "'syslog'"}` ?

**Comment by jappeace** [comment]
Thank you!

The RFC this PR follows: https://github.com/NixOS/rfcs/pull/42
The idea is to only support essential options an upstream program provides, the other options can be set in an easy to use attribute set.

**Review by aszlig — COMMENTED** [review]
As usual: Thanks for revamping the pgsql/mysql modules :-) Commented on a few nitpicks.

File: nixos/modules/services/databases/postgresql.nix [hunk 2]
```diff
@@ -11,23 +11,23 @@ let
       then cfg.package
       else cfg.package.withPackages (_: cfg.extraPlugins);
 
+  toStr = value:
+    if true == value then "yes"
+    else if false == value then "no"
+    else if isString value then "'${value}'"
```
[the remark this answers is not in the document]
Comment (aanderse) [reply nixos/modules/services/databases/postgresql.nix]: Great suggestion.

**Comment by aanderse** [comment]
Feedback addressed.

**Review by aszlig — APPROVED** [review]

**Comment by aanderse** [comment]
@GrahamcOfBorg test postgresql postgresql-wal-receiver

**Comment by aanderse** [comment]
Merge conflict resolved. Anyone keen on merging?

**Comment by rvl** [comment]
It would be kinder to users to have a smooth upgrade path.
Is it possible to deprecate `extraConfig` for one release then remove it in the next release?

**Comment by aanderse** [comment]
> It would be kinder to users to have a smooth upgrade path.
> Is it possible to deprecate `extraConfig` for one release then remove it in the next release?

While I hate making changes that instantly force users to adapt without prior notice, in the case of `extraConfig` it is "worth it", at least in my opinion.

When asked about keeping both a `settings` and `extraConfig` option @Infinisil [had this to say](https://github.com/NixOS/rfcs/pull/42#issuecomment-475873999), which I think is quite relevant:
> I'd rather not if possible, because then we have problem of not being able to merge values set in `extraConfig` with all other values, there are essentially two incompatible ways of specifying configuration then.

Further to this I would argue that migrating from `extraConfig` to `settings` is very trivial so it shouldn't take more than a minute or two.

Have I sufficiently addressed your concerns? I'm happy to continue the discussion and consider this further if not.

**Comment by rvl** [comment]
I suspect that users with existing nixos configurations may resent this as a gratuitous change providing no benefit to them, especially if they need to modify multiple services, or if option merging semantics change. Just my 2c.

---

Related to this PR itself, I have
```
services.postgresql.extraConfig = ''
  session_preload_libraries = 'auto_explain'
  auto_explain.log_min_duration = '5s'
'';
```
Should the postgesql settings option support nested attrsets? E.g.
```
services.postgresql.settings.auto_explain.log_min_duration = "5s";
```

**Comment by aanderse** [comment]
> I suspect that users with existing nixos configurations may resent this as a gratuitous change providing no benefit to them, especially if they need to modify multiple services, or if option merging semantics change. Just my 2c.

I would highly encourage you to express your opinions on this in the https://github.com/NixOS/rfcs/pull/42 thread. The more feedback an RFC gets (constructive criticism included) the better the end result.

> Should the postgesql settings option support nested attrsets? E.g.

I'm not much of a `postgresql` user. I was unaware that such options existed. Documentation should be updated to reflect this and explain values like that should be quoted, like so:
```
services.postgresql.settings."auto_explain.log_min_duration" = "5s";
```

As a `postgresql` user would you have any motivation to contribute to this module by enhancing the documentation?

**Comment by Infinisil** [comment]
I just posted [another comment](https://github.com/NixOS/rfcs/pull/42#issuecomment-691392134) about backwards compatibility in the RFC (saying it's up to the module writer to decide how to handle backwards compatibility). I guess the [original comment](https://github.com/NixOS/rfcs/pull/42#issuecomment-475873999) by me that @aanderse linked is still relevant, but I think I mostly wrote that in relation to *new* NixOS modules. So if you create a new module, it should definitely not have an `extraConfig` if it can instead be `settings`. The RFC was mainly just intended for new modules because I knew backwards compatibility could be a problem.

While I'd love for all existing modules to eventually adapt the `settings` approach, I think backwards compatibility is still important, and for popular modules like postgresql it might make sense to keep supporting `extraConfig` for a while before deprecating it, even though it won't be ideal during that time.

**Comment by aanderse** [comment]
@rvl NixOS is a pretty bleeding edge distro. Unfortunately inconvenient changes happen. I really like the welcoming and collaborative nature of the NixOS community so when I ask: "is this migration really that big of a hassle?" I'm just trying to better understand your perspective.

Here is my perspective: I manage upwards of 40 NixOS machines, each of which has a number of changes required to migrate to `20.09`. Out of all the changes required to migrate to `20.09` I personally see `extraConfig` to `settings` as one of the least impacting and quickest to manage.

So now that we have had this discussion what can you do if you're still not satisfied with the situation? I would be disappointed if you made a PR add `extraConfig` back, but I wouldn't discourage it. If you see this as a breaking which isn't acceptable then we can quickly talk to the module maintainers and release managers to get your PR into `20.09`. I hope you don't feel this way still, but I'll respect your opinion and try to support your efforts the best I can.

**Comment by rvl** [comment]
I think I have already made my perspective clear, thanks.
Please make sure you put something in the 20.09 release notes about this.

**Pull request merged**